codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -308,14 +307,35 @@ class KotlinJmespathExpressionVisitor(
     }
 
     override fun visitSubexpression(expression: Subexpression): VisitedExpression {
-        val left = expression.left.accept(this)
+        val leftName = expression.left.accept(this).identifier
 
-        val ret = when (val right = expression.right) {
-            is FieldExpression -> subfield(right, left.identifier)
+        return when (val right = expression.right) {
+            is FieldExpression -> subfield(right, leftName)
+            is IndexExpression -> index(right, leftName)
+            is Subexpression -> subexpression(right, leftName)
             else -> throw CodegenException("Subexpression type $right is unsupported")
         }
+    }
+
+    private fun subexpression(expression: Subexpression, parentName: String): VisitedExpression {
+        val leftName = when (val left = expression.left) {
+            is FieldExpression -> subfield(left, parentName).identifier
+            is Subexpression -> subexpression(left, parentName).identifier
+            else -> throw CodegenException("Subexpression type $left is unsupported")
```

Comment:
Could `left` also be an `IndexExpression`?

Reply:
I guess it could but it really shouldn't be. The expression would be trying to access an index in nothing i.e. "[2]..."

Reply:
Once all expression visitors are supported I'll change the message to something like `"Subexpression type $left is unsupported as left in $expression"`